### ISSUE TITLE
feat(protocol): reusable ML training and evaluation protocols

### ISSUE THREAD

**Issue by junwen94** [stfc/goldilocks-ml#1]
Establish reusable ML training and evaluation protocols: splits, metrics, baselines and reproducible runs, shared by every model in this repository.

From the v0.1 roadmap deliverables (ml, v0.1c). Filed so the v0.1 milestone reflects its real scope; scope and design still to be worked out.

**Comment by junwen94** [comment]
## Proposed design: executable training protocols

The protocol should be executable and auditable, not only a README or notebook. A colleague should be able to start from an immutable dataset snapshot, run one documented command, and obtain a self-describing run bundle that can be evaluated and published with the model.

The historical `stfc/goldilocks_kpoints` repository is a scientific reference, not a runtime dependency. Its relative paths, mutable CSV/LMDB outputs, notebook-based conformal calibration, local MLflow state, and platform-specific dependency setup should not be copied as the new public contract.

### Scope and ownership

Issue #1 owns the shared protocol, split, evaluation, and run-bundle machinery. Model implementations remain separate:

- #2: k-mesh/QRF training protocol and model implementation;
- #3: CGCNN metallicity training protocol and model implementation.

`goldilocks-data` owns immutable dataset snapshots. `goldilocks-ml` verifies and consumes them. `goldilocks-core` only consumes released inference artifacts.

### User workflow

```bash
uv sync --group train

uv run goldilocks-train validate \
  protocols/kmesh/qrf95.toml \
  --dataset /path/to/dataset-snapshot

uv run goldilocks-train run \
  protocols/kmesh/qrf95.toml \
  --dataset /path/to/dataset-snapshot \
  --output local_runs/qrf95-v1
```

`validate` performs no training. `run` validates again, creates the split manifest, fits preprocessing on the training split only, trains the configured baseline and model, evaluates validation/calibration/test data according to the protocol, and writes one run bundle. There is no notebook-only step.

### Versioned protocol

Store reviewed protocols under `protocols/<task>/<model>.toml`. Use stable built-in trainer names rather than arbitrary Python import strings.

```toml
schema_version = 1
id = "kmesh-qrf95-v1"
task = "regression"
trainer = "quantile_random_forest"

[dataset]
record_id = "PSDI-record-id"
snapshot_version = "version"
manifest_sha256 = "sha256"
sample_id = "sample_id"
target = "k_distance"

[split]
method = "group"
group_column = "structure_group_id"
train = 0.7
validation = 0.1
calibration = 0.1
test = 0.1
seed = 42

[features]
schema = "qrf_comp_struct_soap_lattice_metal"

[model]
seed = 42
n_estimators = 100
quantiles = [0.05, 0.5, 0.95]

[evaluation]
primary_metric = "mae"
metrics = ["mae", "rmse", "r2", "pinball_loss", "interval_coverage", "mean_interval_width"]
baseline = "train_median"
```

The schema must also support classification protocols. A metallicity protocol should use stratified group splitting where possible, select any decision threshold on validation data only, and report MCC, balanced accuracy, precision, recall, F1, confusion matrix, ROC-AUC, and PR-AUC when probabilities exist.

### Dataset and split contract

- Require a dataset manifest with an immutable record/version and SHA-256.
- Require stable sample IDs; never split by dataframe row position.
- Write `splits.csv` with `sample_id,split` and reuse it on reruns.
- Validate disjoint train/validation/calibration/test IDs.
- Support an explicit grouping column for structure, composition, prototype, or calculation-family leakage control.
- Fit scalers, imputers, feature selection, and learned preprocessing on training data only.
- Keep calibration separate from validation and test when conformal correction is requested.
- Test data is evaluated once after model/configuration selection; it is never used for early stopping, threshold selection, calibration, or model choice.

Random splitting remains available only as an explicit protocol choice. Historical behavior should be labelled `legacy_reproduction`; it must not silently become the default scientific claim.

### Run bundle

Write under ignored `local_runs/`:

```text
local_runs/<run-id>/
├── run.json                 # run id, timestamps, git commit, status
├── protocol.toml            # fully resolved protocol
├── dataset.json             # source record/version/hash
├── environment.json         # Python, package, lock hash, hardware facts
├── splits.csv               # stable sample-to-split assignment
├── metrics.json             # baseline and model metrics by split/slice
├── predictions.csv          # sample id, truth, prediction/score/interval
├── model/                   # model-specific artifacts
└── manifest.json            # size and SHA-256 for every publishable file
```

Do not require a remote MLflow service. Optional experiment tracking may consume this bundle later, but the filesystem bundle is authoritative.

The run is scientifically reproducible when the same dataset snapshot, protocol, code commit, and locked environment are available. Byte-identical model artifacts are promised only when the selected trainer and hardware path are documented as deterministic.

### Publication

- Track protocol TOML and human documentation in Git and render them on the existing MkDocs site.
- Publish the resolved protocol, dataset identity, split manifest, metrics, and run manifest alongside a released model where licensing permits.
- Reuse `goldilocks-psdi checksum` and the existing draft-upload workflow; training code must not contain a second PSDI client.
- Do not claim that the already published QRF95 or CGCNN artifacts are exactly reproducible until their historical dataset snapshot, split assignment, complete configuration, and code revision are recovered.

### Package structure

```text
protocols/                    # reviewed TOML protocols
src/goldilocks_ml/
├── protocol.py              # schema loading and validation
├── datasets.py              # snapshot and column contract
├── splitting.py             # deterministic split manifests/leakage checks
├── evaluation.py            # shared metrics and baselines
├── runs.py                  # run bundle and checksums
└── train_cli.py             # thin validate/run CLI
tests/protocols/              # synthetic, offline fixtures
docs/training-protocols.md
```

Model-specific feature extraction and trainers belong in scientific task modules added by #2 and #3, not in generic helpers.

### Phases

#### Phase 1: protocol and dataset contracts

- [ ] Define the TOML schema and immutable dataset snapshot contract.
- [ ] Implement strict loading/validation with unknown-field rejection.
- [ ] Add a documented synthetic protocol fixture.

Verification: malformed protocols, missing hashes/columns, and incompatible task settings fail before training.

#### Phase 2: split and evaluation primitives

- [ ] Generate/reload stable split manifests from sample IDs.
- [ ] Add group leakage checks and train-only preprocessing boundaries.
- [ ] Implement regression/classification metrics and simple train-derived baselines.

Verification: synthetic tests prove deterministic splits, no overlap/group leakage, correct metrics, and no preprocessing fit outside train.

#### Phase 3: run CLI and bundle

- [ ] Add `goldilocks-train validate` and `goldilocks-train run`.
- [ ] Record resolved protocol, dataset, environment, predictions, metrics, and artifact hashes.
- [ ] Add one CPU-only end-to-end fixture trainer so CI exercises the complete workflow without private data or network access.

Verification: a clean checkout produces the expected bundle twice with identical split and manifest identities.

#### Phase 4: documentation and model handoff

- [ ] Publish the colleague-facing workflow on GitHub Pages.
- [ ] Document reproduction versus compatible retraining and deterministic limitations.
- [ ] Define the trainer interface consumed by #2 and #3.

Verification: docs commands are tested and the wheel contains the required schema/template resources.

### Acceptance criteria

- [ ] One command validates a protocol and dataset without training or network access.
- [ ] One command performs a complete synthetic CPU run and produces the documented bundle.
- [ ] Split IDs, dataset identity, resolved configuration, environment, metrics, predictions, and artifact hashes are recorded.
- [ ] Test-set isolation and train-only preprocessing are enforced by tests.
- [ ] No private data, large artifacts, notebooks, credentials, or MLflow state are required by CI.
- [ ] QRF and CGCNN can adopt the shared interface in #2 and #3 without duplicating split/evaluation/run-bundle code.

### Dependencies and open inputs

- The first real model protocol depends on an immutable `goldilocks-data` snapshot; that repository is currently empty and its data protocol is tracked in `stfc/goldilocks-data#1`.
- Recommended first vertical slice after the shared protocol: QRF, because it is CPU-trainable and exercises regression, quantiles, calibration, and model serialization before adding the heavier PyTorch/graph stack.
- The exact dataset record/version/hash and leakage grouping field must be supplied by the data snapshot, not invented in this repository.

---
Written by an agent on behalf of Junwen Yin.
